include/Dialect/Secret/IR/SecretOps.td

```diff
@@ -0,0 +1,171 @@
+    Examples:
+
+    ```
+    %Y = secret.reveal %secret_value : !secret.secret<i32> -> i32
+    ```
+  }];
+
+  let arguments = (ins Secret:$input);
+  let results = (outs AnyType:$cleartext);
+  let assemblyFormat = "$input attr-dict `:` type($input) `->` type($cleartext)";
+
+  let builders = [
+    // Builder to infer output type from the input type
+    OpBuilder<(ins "Value":$secret)>
+  ];
+}
+
+
+def Secret_YieldOp : Secret_Op<"yield", [Pure, ReturnLike, Terminator]>,
+  Arguments<(ins Variadic<AnyType>:$values)> {
+  let summary = "Secret yield operation";
+  let description = [{
+    `secret.yield` is a special terminator operation for blocks inside regions
+    in `secret` generic ops. It returns the cleartext value of the
+    corresponding private computation to the immediately enclosing `secret`
+    generic op.
+  }];
+  let builders = [OpBuilder<(ins), [{ /* nothing to do */ }]>];
+  let hasCustomAssemblyFormat = 1;
+
+  // TODO(https://github.com/google/heir/issues/108): add a verifier that
+  // ensures the yield op is inside a generic op, that the return type of the
+  // yield op matches the enclosing generic, and any other similar
+  // verifications that linalg.generic has that would be applicable here. Maybe
+  // there are some generic interfaces that produce these verifiers for free,
+  // like SingleBlockImplicitTerminator?
+  // let hasVerifier = 1;
+}
+
+def Secret_GenericOp : Secret_Op<"generic", [
+       SingleBlockImplicitTerminator<"YieldOp">]> {
+  let summary = "Lift a plaintext computation to operate on secrets.";
+  let description = [{
+    `secret.generic` lifts a plaintext computation to operate on one or more
+    secrets. The lifted computation is represented as a region containing a
+    single block terminated by `secret.yield`. The arguments of the `secret.generic`
+    may include one or more `!secret.secret` types. The arguments of the block
+    in the op's body correspond to the underlying plaintext types of the secrets.
+
+    Basic examples:
+
+    Add two secret integers together
+
+    ```
+    %Z = secret.generic ins(%X, %Y : !secret.secret<i32>, !secret.secret<i32>) {
+      ^bb0(%x: i32, %y: i32) :
+        %z = arith.addi %x, %y: i32
+        secret.yield %z : i32
+      } -> (!secret.secret<i32>)
+    ```
+
+    Add a secret value with a plaintext value. I.e., not all arguments to the
+    op need be secret.
+
+    ```
+    %Z = secret.generic ins(%X, %Y : i32, !secret.secret<i32>) {
+      ^bb0(%x: i32, %y: i32) :
+        %z = arith.addi %x, %y: i32
+        secret.yield %z : i32
+      } -> (!secret.secret<i32>)
+    ```
+
+    A `secret.generic` op also has a corresponding `library_name` attribute
+    that provides semantic information about what op is computed inside the
+    region. This can be used by lowering passes to target specific
+    instructions in lower level dialects that support them. For example, an FHE
+    dialect may have special support for an encrypted matmul op.
```

Comment:
I have a feeling the `library_name` concept is a bit too low-level for such a high-level concept. 
I'd expect the `fhe`/`enc_linalg`/`enc_arith` level dialect's `generic` to have this, but not `secret` (if we are indeed adding it "on top" of the encrypted arith abstraction)

Reply:
I would have thought that `secret` is actually a higher level than `fhe`?

Reply:
I think what Alex is saying makes sense, in that a program using `secret` would be wrapping a named op already (e.g., `secret` wrapping `tosa.matmul`) and that in the lowering to lower-level dialects you'd take the wrapped op and use it for `library_name` on a region of ops in `fhe_math` that implements something manually.

Reply:
AH, I see. I'm in agreement, that's right.